src/CommunityToolkit.Maui.Core/Essentials/FileSaver/FileSaverImplementation.android.cs

```diff
@@ -84,11 +84,34 @@ static AndroidUri EnsurePhysicalPath(AndroidUri? uri)
 
 	static async Task<string> SaveDocument(AndroidUri uri, Stream stream, CancellationToken cancellationToken)
 	{
+		if (stream.CanSeek)
+		{
+			stream.Seek(0, SeekOrigin.Begin);
+		}
+
 		using var parcelFileDescriptor = Application.Context.ContentResolver?.OpenFileDescriptor(uri, "wt");
 		using var fileOutputStream = new FileOutputStream(parcelFileDescriptor?.FileDescriptor);
-		await using var memoryStream = new MemoryStream();
-		await stream.CopyToAsync(memoryStream, cancellationToken).ConfigureAwait(false);
-		await fileOutputStream.WriteAsync(memoryStream.ToArray()).WaitAsync(cancellationToken).ConfigureAwait(false);
+		var buffer = ArrayPool<byte>.Shared.Rent(4096);
+
+		try
+		{
+			int bytesRead;
+			long totalRead = 0;
+			while ((bytesRead = await stream.ReadAsync(buffer, cancellationToken).ConfigureAwait(false)) > 0)
+			{
+				await fileOutputStream.WriteAsync(buffer, 0, bytesRead).WaitAsync(cancellationToken).ConfigureAwait(false);
+				totalRead += bytesRead;
+			}
+
+			if (fileOutputStream.Channel is not null)
+			{
+				await fileOutputStream.Channel.TruncateAsync(totalRead).WaitAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
What is the purpose of truncating here? Is it to ensure that we have flushed everything out to disk?

Reply:
Yes. 
Assume you have a file size 1000 bytes. Then you try overwriting it with the 500 bytes. Without truncate the file size remains 1000 bytes

Reply:
Thanks, I hadn't considered the scenario where we are overwriting an existing file.